cmd/storage-rest-server.go

```diff
@@ -353,6 +354,55 @@ func (s *storageRESTServer) ReadFileStreamHandler(w http.ResponseWriter, r *http
 	w.(http.Flusher).Flush()
 }
 
+func readMetadata(s StorageAPI, volume, entry, leafFile string) FileInfo {
+	buf, err := s.ReadAll(volume, pathJoin(entry, leafFile))
+	if err != nil {
+		return FileInfo{}
+	}
+	m, err := xlMetaV1UnmarshalJSON(context.Background(), buf)
+	if err != nil {
+		return FileInfo{}
+	}
+	return FileInfo{
+		Volume:   volume,
+		Name:     entry,
+		ModTime:  m.Stat.ModTime,
+		Size:     m.Stat.Size,
+		Metadata: m.Meta,
+		Parts:    m.Parts,
+	}
```

Comment:
instead of FileInfo you can use different struct WalkEntryInfo

```
type struct WalkEntryInfo{
Name: string
ModTime: time.Time
Size:
StorageClass:
}
```
and use gjson to query only required entries.

Reply:
also in xlMetaV1UnmarshalJSON() gjson based json unmarshal is removed? how come?

Reply:
gjson based is not needed anymore since jsoniter is as fast @krishnasrinivas and only 

```
type struct WalkEntryInfo{
Name: string
ModTime: time.Time
Size:
StorageClass:
}
``` 

is not possible anymore, we need the entire set of Meta and Parts information to be captured as well to work with compressed and encrypted objects. So gjson is not useful when you are reading the entire file anyways.

Reply:
> is not possible anymore, we need the entire set of Meta and Parts information to be captured as well to work with compressed and encrypted objects

posix layer can unmarshal the xl.json and peek into the metadata providing the actual size and whatever info S3's list object needs.

Reply:
No, it cannot @krishnasrinivas it requires entire compression/encryption related logic going into POSIX which is not correct since they are only available at the handler layer. Spilling code like that will make it complicated.

Reply:
I see.
won't interpreting `X-Minio-Internal-actual-size` be good enough?
what compression/encryption logic will spill over to posix if we need to do it?

Reply:
No that's just for compression @krishnasrinivas we need to be able to decrypt the ETag, and also use sio library to find the right decrypted size as well.